code/game/machinery/computer/communications.dm

```diff
@@ -500,6 +501,418 @@
+						log_game("A distress beacon requested by [key_name_admin(usr)] was automatically sent due to not receiving an answer within 60 seconds.")
+						message_admins("A distress beacon requested by [ADMIN_TPMONTY(usr)] was automatically sent due to not receiving an answer within 60 seconds.")
+						return TRUE
+			else
+				state = STATE_DISTRESS */
+
+		if("messagelist")
+			currmsg = 0
+			state = STATE_MESSAGELIST
+
+		if("viewmessage")
+			state = STATE_VIEWMESSAGE
+			if(!currmsg)
+				if(href_list["message-num"])
+					currmsg = text2num(href_list["message-num"])
+				else
+					state = STATE_MESSAGELIST
+
+		if("delmessage")
+			state = (currmsg) ? STATE_DELMESSAGE : STATE_MESSAGELIST
+
+		if("delmessage2")
+			if(authenticated)
+				if(currmsg)
+					var/title = messagetitle[currmsg]
+					var/text  = messagetext[currmsg]
+					messagetitle.Remove(title)
+					messagetext.Remove(text)
+					if(currmsg == aicurrmsg)
+						aicurrmsg = 0
+					currmsg = 0
+				state = STATE_MESSAGELIST
+			else
+				state = STATE_VIEWMESSAGE
+
+
+		if("status")
+			state = STATE_STATUSDISPLAY
+
+		// Status display stuff
+		if("setstat")
+			switch(href_list["statdisp"])
+				if("message")
+					post_status("message", stat_msg1, stat_msg2)
+				if("alert")
+					post_status("alert", href_list["alert"])
+				else
+					post_status(href_list["statdisp"])
+
+		if("setmsg1")
+			stat_msg1 = reject_bad_text(trim(copytext(sanitize(input("Line 1", "Enter Message Text", stat_msg1) as text|null), 1, 40)), 40)
+			updateDialog()
+
+		if("setmsg2")
+			stat_msg2 = reject_bad_text(trim(copytext(sanitize(input("Line 2", "Enter Message Text", stat_msg2) as text|null), 1, 40)), 40)
+			updateDialog()
+
+		if("messageTGMC")
+			if(authenticated == 2)
+				if(world.time < cooldown_central + COOLDOWN_COMM_CENTRAL)
+					to_chat(usr, "<span class='warning'>Arrays recycling.  Please stand by.</span>")
+					return FALSE
+
+				var/input = stripped_input(usr, "Please choose a message to transmit to the TGMC High Command.  Please be aware that this process is very expensive, and abuse will lead to termination.  Transmission does not guarantee a response. There is a small delay before you may send another message. Be clear and concise.", "To abort, send an empty message.", "")
+				if(!input || !(usr in view(1,src)) || authenticated != 2 || world.time < cooldown_central + COOLDOWN_COMM_CENTRAL)
+					return FALSE
+
+
+				Centcomm_announce(input, usr)
+				to_chat(usr, "<span class='notice'>Message transmitted.</span>")
+				usr.log_talk(input, LOG_SAY, tag="TGMC announcement")
+				cooldown_central = world.time
+
+		if("securitylevel")
+			tmp_alertlevel = text2num( href_list["newalertlevel"] )
+			if(!tmp_alertlevel)
+				tmp_alertlevel = 0
+			state = STATE_CONFIRM_LEVEL
+
+		if("changeseclevel")
+			state = STATE_ALERT_LEVEL
+
+		else return FALSE
+
+	updateUsrDialog()
+
+/obj/machinery/computer/communications/cmo/attack_hand(var/mob/user as mob)
+	if(..())
```

Comment:
Calling upper here and adding another snowflake chcek for area is a bad solution, either don't call upper and have each check separate or make this into a variable (that would be better)